code/modules/power/turbine/turbine.dm

```diff
@@ -0,0 +1,557 @@
+#define TURBINE_WORK_CONVERSION_MULTIPLIER 0.01
+///Multiplier when converting gas work back into heat
+#define TURBINE_HEAT_CONVERSION_MULTIPLIER 0.005
+///Amount of energy removed from the work done by the stator due to the consumption from the compressor working on the gases
+#define TURBINE_COMPRESSOR_STATOR_INTERACTION_MULTIPLIER 0.15
+
+/obj/machinery/power/turbine
+	density = TRUE
+	resistance_flags = FIRE_PROOF
+	can_atmos_pass = ATMOS_PASS_DENSITY
+
+	///Theoretical volume of gas that's moving through the turbine, it expands the further it goes
+	var/gas_theoretical_volume = 0
+	///Stores the turf thermal conductivity to restore it later
+	var/our_turf_thermal_conductivity
+	///Checks if the machine is processing or not
+	var/active = FALSE
+	///The parts can be registered on the main one only when their panel is closed
+	var/can_connect = TRUE
+
+	///Overlay for panel_open
+	var/open_overlay //#TODO: get the overlay done
+	///Overlay for machine activation
+	var/on_overlay //#TODO: get the overlay done
+
+	///Reference to our turbine part
+	var/obj/item/turbine_parts/installed_part
+	///Path of the turbine part we can install
+	var/part_path
+
+/obj/machinery/power/turbine/Initialize(mapload)
+	. = ..()
+
+	if(part_path)
+		installed_part = new part_path(src)
+
+	var/turf/our_turf = get_turf(src)
+	if(our_turf.thermal_conductivity != 0 && isopenturf(our_turf))
+		our_turf_thermal_conductivity = our_turf.thermal_conductivity
+		our_turf.thermal_conductivity = 0
+
+/obj/machinery/power/turbine/Destroy()
+	var/turf/our_turf = get_turf(src)
+	if(our_turf.thermal_conductivity == 0 && isopenturf(our_turf))
+		our_turf.thermal_conductivity = our_turf_thermal_conductivity
+
+	if(installed_part)
+		QDEL_NULL(installed_part)
+
+	return ..()
+
+/obj/machinery/power/turbine/screwdriver_act(mob/living/user, obj/item/tool)
+	if(active)
+		to_chat(user, "You can't open [src] while it's on!")
+		return TOOL_ACT_TOOLTYPE_SUCCESS
+	if(!anchored)
+		to_chat(user, span_notice("Anchor [src] first!"))
+		return TOOL_ACT_TOOLTYPE_SUCCESS
+
+	tool.play_tool_sound(src, 50)
+	panel_open = !panel_open
+	if(panel_open)
+		disable_parts(user)
+	else
+		enable_parts(user)
+	var/descriptor = panel_open ? "open" : "close"
+	balloon_alert(user, "you [descriptor] the maintenance hatch of [src]")
+	update_appearance()
+	return TOOL_ACT_TOOLTYPE_SUCCESS
+
+/obj/machinery/power/turbine/update_overlays()
+	. = ..()
+	if(panel_open)
+		. += open_overlay
+	if(active)
+		. += on_overlay
+
+/obj/machinery/power/turbine/wrench_act(mob/living/user, obj/item/tool)
+	return default_change_direction_wrench(user, tool)
+
+/obj/machinery/power/turbine/crowbar_act(mob/living/user, obj/item/tool)
+	return default_deconstruction_crowbar(tool)
+
+/obj/machinery/power/turbine/proc/enable_parts(mob/user)
+	can_connect = TRUE
+
+/obj/machinery/power/turbine/proc/disable_parts(mob/user)
+	can_connect = FALSE
+
+/obj/machinery/power/turbine/Moved(atom/OldLoc, Dir)
```

Comment:
isn't it oldLoc?

Reply:
Dunno tbh it autocompleted like that

Reply:
![image](https://user-images.githubusercontent.com/58055496/159806215-03165d5d-341c-4728-9940-5636f585363b.png)
fuckin